### PR TITLE
Scheduling enclaves in the C target

### PR DESCRIPTION
**Pull request by erlingrj**
This PR adds the first version of the implementation of scheduling enclaves in the C target. It is paired with this PR in reactor-c: https://github.com/lf-lang/reactor-c/pull/242

### CODE REVIEW

File: core/src/main/java/org/lflang/generator/c/CEnclaveGenerator.java [hunk 1]
```diff
@@ -0,0 +1,390 @@
...
+import java.util.stream.Collectors;
+
+import org.lflang.ErrorReporter;
+import org.lflang.TargetConfig;
+import org.lflang.TimeValue;
+import org.lflang.generator.CodeBuilder;
+import org.lflang.generator.ReactorInstance;
+import org.lflang.generator.c.CEnclaveGraph.EnclaveConnection;
+
+// FIXME: This file should be renamed to CEnclaveGenerator or something.
```
Comment (edwardalee) [inline core/src/main/java/org/lflang/generator/c/CEnclaveGenerator.java:15]: This FIXME seems to be fixed?

File: core/src/main/java/org/lflang/generator/c/CEnclaveGenerator.java [hunk 2]
```diff
@@ -0,0 +1,390 @@
...
+  public CEnclaveGenerator(
+      ReactorInstance main, TargetConfig targetConfig, String lfModuleName, ErrorReporter errorReporter) {
+    this.enclaves = CUtil.getEnclaves(main);
+    this.targetConfig = targetConfig;
+    this.lfModuleName = lfModuleName;
+    this.errorReporter = errorReporter;
+    this.enclaveGraph = new CEnclaveGraph(this.enclaves);
+
+    // FIXME: This is not really the best place for the cycle detection. The problem is that the EnclaveGraph
+    //  assumes that is being run on the post AST transformation graph. So cannot be used in the ValidatorCheck (which is pre-
```
Comment (edwardalee) [inline core/src/main/java/org/lflang/generator/c/CEnclaveGenerator.java:36]:
```suggestion
    //  assumes that is being run on the post AST transformation graph. So cannot be used in the ValidatorCheck 
    // (which is pre-AST transformations).
```

File: core/src/main/java/org/lflang/generator/c/CEnclaveGraph.java [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+import java.util.Set;
+import java.util.Stack;
+
+import org.lflang.TimeValue;
+import org.lflang.generator.PortInstance;
+import org.lflang.generator.ReactorInstance;
+import org.lflang.generator.RuntimeRange;
+import org.lflang.generator.SendRange;
+
+public class CEnclaveGraph {
```
Comment (edwardalee) [inline core/src/main/java/org/lflang/generator/c/CEnclaveGraph.java:15]: Why doesn't this class use the existing graph classes?

File: core/src/main/java/org/lflang/validation/AttributeSpec.java [hunk 4]
```diff
@@ -213,10 +214,12 @@ enum AttrParamType {
     ATTRIBUTE_SPECS_BY_NAME.put(
         "side",
         new AttributeSpec(List.of(new AttrParamSpec(VALUE_ATTR, AttrParamType.STRING, false))));
-    // @enclave(each=boolean)
+    // @enclave(workers=inteach=boolean)
```
Comment (edwardalee) [inline core/src/main/java/org/lflang/validation/AttributeSpec.java:217]: Should this be:
```suggestion
    // @enclave(workers=int, each=boolean)
```
What is the meaning of the each parameter?

File: core/src/main/java/org/lflang/validation/LFValidator.java [hunk 5]
```diff
@@ -411,6 +416,88 @@ public void checkDeadline(Deadline deadline) {
...
+      }
+      for (Output output : encDef.getOutputs()) {
+        if (output.getWidthSpec() != null) {
+          error("Enclaves with multiports not supported in the C target", Literals.WIDTH_SPEC__TERMS);
+        }
+      }
+
+      // 3. Disallow enclaves inside modes
+      EObject container = inst.eContainer();
+      if (container instanceof Mode) {
```
Comment (edwardalee) [inline core/src/main/java/org/lflang/validation/LFValidator.java:442]: Should this be checking if _any_ container is a mode?

**Review by edwardalee — COMMENTED** [review]
Wow, this is a big PR. A real tour-de-force.  I've made some suggestions and raised some questions, but I didn't review closely all the code.

**Comment by erlingrj** [comment]
Thanks for feedback Edward. I will address this. The PR is still a draft and I think I should do a proper pass through it and fix documentation and etc before anyone looks at it seriously

File: core/src/main/java/org/lflang/AttributeUtils.java [hunk 6]
```diff
@@ -255,16 +255,32 @@ public static boolean isEnclave(Instantiation node) {
...
+     }
+    }
+    return 1; // Not specified
+  }
+
   /**
-   * Annotate @{code node} with enclave @attribute
-   *
-   * @param node
+   * Move the enclave attribute from source inst to target inst.
```
[the remark this answers is not in the document]
Comment (erlingrj) [reply core/src/main/java/org/lflang/AttributeUtils.java]: Thanks, fixed

File: core/src/main/java/org/lflang/ast/ASTUtils.java [hunk 7]
```diff
@@ -1853,4 +1858,21 @@ public static void addReactionAttribute(Reaction reaction, String name) {
...
   }
+
+  public static List<Instantiation> getEnclaves(Reactor top) {
+    List<Instantiation> enclaves = new ArrayList<>();
+    Queue<Reactor> queue = new LinkedList<>();
+    queue.add(top);
+
+    while (!queue.isEmpty()) {
+      Reactor inst = queue.poll();
+      for (Instantiation child: inst.getInstantiations()) {
```
[the remark this answers is not in the document]
Comment (erlingrj) [reply core/src/main/java/org/lflang/ast/ASTUtils.java]: Thanks, fixed.

File: core/src/main/java/org/lflang/generator/ReactionInstanceGraph.java [hunk 8]
```diff
@@ -101,19 +103,36 @@ public void rebuildAndAssignDeadlines() {
...
+      var breadth = breadthMap.get(enclave);
+      if (breadth != null) {
+        res.add(breadth);
+      } else {
+        res.add(0);
+      }
+    }
+    return res.toArray(new Integer[0]);
+  }
   /*
```
[the remark this answers is not in the document]
Comment (erlingrj) [reply core/src/main/java/org/lflang/generator/ReactionInstanceGraph.java]: Fixed

File: core/src/main/java/org/lflang/generator/c/CEnclaveGenerator.java (hunk 1 above)
Comment (erlingrj) [reply core/src/main/java/org/lflang/generator/c/CEnclaveGenerator.java]: Yes

File: core/src/main/java/org/lflang/generator/c/CEnclaveGenerator.java (hunk 2 above)
Comment (erlingrj) [reply core/src/main/java/org/lflang/generator/c/CEnclaveGenerator.java]: Fixed

**Comment by erlingrj** [comment]
@lhstrh and @edwardalee, I am tagging you for a review now. We still have a test failure on macOS with coordinated shutdown of enclaves which I will look at in ~10 days. I have made a pass over all the changes and tried to bring the docs up to date. I kept my custom graph implementation due to the need to annotate edges, not nodes as is done in the other graphs. I will be mostly offline the next 8 days.

**Comment by erlingrj** [comment]
Closing this as we are pursuing a different AST strategy. See #2104

**Pull request closed without merging**